examples/hhl.py

```diff
@@ -0,0 +1,414 @@
+           │         │QFT^-1│    │      │      │      │ │QFT│         │
+(2, 0): ─H─┼─────@───│      │──X─@────X─@────X─@────X─@─│   │─@───────┼─H────
+           │     │   └──────┘                           └───┘ │       │
+(3, 0): ───e^iAt─e^2iAt───────────────────────────────────────e^-2iAt─e^-iAt─
+
+Note: QFT in the above diagram omits swaps, which are included implicitly by
+reversing qubit order for phase kickbacks.
+"""
+
+import math
+import cirq
+import numpy as np
+
+
+class PhaseEstimation(cirq.MultiQubitGate):
+    """
+    A gate for Quantum Phase Estimation.
+
+    U is the unitary matrix whose phases will be estimated.
+    The last qubit stores the eigenvector; all other qubits store the
+    estimated phase, in big-endian.
+    """
+
+    def __init__(self, num_qubits, U, invert=False):
+        super().__init__(num_qubits)
+        self.U = U
+        self.invert = invert
+
+    def _decompose_(self, qubits):
+        if self.invert:
+            qubits = list(qubits)
+            yield Qft(self._num_qubits-1)(*qubits[:-1])
+            yield (PhaseKickback(self.num_qubits(), self.U)**-1)(*qubits)
+            yield cirq.H.on_each(qubits[:-1])
+        else:
+            qubits = list(qubits)
+            yield cirq.H.on_each(qubits[:-1])
+            yield PhaseKickback(self.num_qubits(), self.U)(*qubits)
+            yield (Qft(self._num_qubits-1)**-1)(*qubits[:-1])
+
+    def __pow__(self, exponent):
+        if exponent == -1:
+            return PhaseEstimation(self.num_qubits(), self.U, invert=True)
+        return NotImplemented
+
+
+class HamiltonianSimulation(cirq.EigenGate, cirq.SingleQubitGate):
+    """
+    A gate that represents e^iAt.
+
+    This EigenGate + np.linalg.eigh() implementation is temporary.
+    Actual implementation will use linear operators.
+    """
+
+    def __init__(self, A, t, exponent=1.0):
+        cirq.SingleQubitGate.__init__(self)
+        cirq.EigenGate.__init__(self, exponent=exponent)
+        self.A = A
+        self.t = t
+        ws, vs = np.linalg.eigh(A)
+        self.eigen_components = []
+        for w, v in zip(ws, vs.T):
+            theta = w*t / math.pi
+            P = np.outer(v, v)
+            self.eigen_components.append((theta, P))
+
+    def _with_exponent(self, exponent):
+        return HamiltonianSimulation(self.A, self.t, exponent)
+
+    def _eigen_components(self):
+        return self.eigen_components
+
+
+class PhaseKickback(cirq.MultiQubitGate):
+    """
+    A gate for the phase kickback stage of Quantum Phase Estimation.
+
+    It consists of a series of controlled e^iAt gates with the memory qubit as
+    the target and each register qubit as the control, raised
+    to the power of 2 based on the qubit index.
+    """
+
+    def __init__(self, num_qubits, U, invert=False):
+        super(PhaseKickback, self).__init__(num_qubits)
+        self.U = U
+        self.invert = invert
+
+    def _decompose_(self, qubits):
+        if self.invert:
```

Comment:
Once https://github.com/quantumlib/Cirq/pull/1407 is merged, you can remove this inverse stuff.